File: fetchr/run/drift_monitor.py
from __future__ import annotations

from typing import Any, Dict

import numpy as np
import pandas as pd


_KEY_FIELDS = [
    "status",
    "error",
    "method",
    "disagg_method_used",
    "auto_selection_reason",
    "auto_selection_strategy",
    "profile_name",
    "series_kind",
    "bootstrap_method",
    "indicator_preprocess_mode",
]

_NUMERIC_FIELDS = [
    "bootstrap_k_step_selected",
    "auto_selection_score_r2",
    "rho",
    "constraint_benchmark_abs_error",
]

_HIGH_SEVERITY_KEYS = {"status", "method", "disagg_method_used"}


def _safe_scalar(value: Any) -> Any:
    if value is None:
        return None
    if isinstance(value, (np.floating, np.integer)):
        value = value.item()
    if isinstance(value, float) and not np.isfinite(value):
        return None
    if isinstance(value, str):
        text = value.strip()
        if not text:
            return None
        as_float = _safe_float(text)
        if as_float is not None:
            return as_float
        return text
    return value


def _safe_float(value: Any) -> float | None:
    if value is None:
        return None
    try:
        v = float(value)
    except Exception:
        return None
    if not np.isfinite(v):
        return None
    return float(v)


def _row_diff(prev: Dict[str, Any], cur: Dict[str, Any], *, score_delta_warn: float) -> Dict[str, Any]:
    changed_keys: Dict[str, Dict[str, Any]] = {}
    for key in _KEY_FIELDS:
        pv = _safe_scalar(prev.get(key))
        cv = _safe_scalar(cur.get(key))
        if pv != cv:
            changed_keys[key] = {"previous": pv, "current": cv}

    numeric_changes: Dict[str, Dict[str, Any]] = {}
    score_delta_abs = 0.0
    for key in _NUMERIC_FIELDS:
        pv = _safe_float(prev.get(key))
        cv = _safe_float(cur.get(key))
        if pv is None and cv is None:
            continue
        if pv is None or cv is None or abs(cv - pv) > 1e-12:
            delta = None if pv is None or cv is None else float(cv - pv)
            numeric_changes[key] = {"previous": pv, "current": cv, "delta": delta}
            if key == "auto_selection_score_r2" and delta is not None:
                score_delta_abs = abs(delta)

    severity = "none"
    if any(k in _HIGH_SEVERITY_KEYS for k in changed_keys):
        severity = "high"
    elif changed_keys or numeric_changes:
        severity = "medium"
    if score_delta_abs >= float(score_delta_warn):
        severity = "high"

    return {
        "changed_keys": changed_keys,
        "numeric_changes": numeric_changes,
        "severity": severity,
        "score_delta_abs": score_delta_abs,
    }


def _rows_by_name(df: pd.DataFrame) -> tuple[Dict[str, Dict[str, Any]], list[str]]:
    rows: Dict[str, Dict[str, Any]] = {}
    duplicates: list[str] = []
    for _, row in df.iterrows():
        name = str(row["name"])
        if name in rows:
            duplicates.append(name)
        rows[name] = dict(row)
    return rows, sorted(set(duplicates))


def build_interpolation_drift_report(
    *,
    current_summary: pd.DataFrame,
    previous_summary: pd.DataFrame | None,
    score_delta_warn: float = 0.05,
) -> Dict[str, Any]:
    cur_df = current_summary.copy()
    if "name" not in cur_df.columns:
        raise ValueError("current_summary must include 'name' column")
    cur_df = cur_df.copy()
    cur_df["name"] = cur_df["name"].astype(str)
    cur_rows, duplicate_names_current = _rows_by_name(cur_df)

    if previous_summary is None or previous_summary.empty:
        return {
            "status": "baseline_initialized",
            "current_count": int(len(cur_rows)),
            "previous_count": 0,
            "added_series": sorted(cur_rows.keys()),
            "removed_series": [],
            "changed_series": [],
            "duplicate_names_current": duplicate_names_current,
            "duplicate_names_previous": [],
            "high_severity_count": int(len(duplicate_names_current)),
        }

    prev_df = previous_summary.copy()
    if "name" not in prev_df.columns:
        return {
            "status": "previous_summary_invalid",
            "current_count": int(len(cur_rows)),
            "previous_count": int(len(prev_df)),
            "added_series": sorted(cur_rows.keys()),
            "removed_series": [],
            "changed_series": [],
            "duplicate_names_current": duplicate_names_current,
            "duplicate_names_previous": [],
            "high_severity_count": int(len(duplicate_names_current)),
        }
    prev_df["name"] = prev_df["name"].astype(str)
    prev_rows, duplicate_names_previous = _rows_by_name(prev_df)

    cur_names = set(cur_rows.keys())
    prev_names = set(prev_rows.keys())
    added = sorted(cur_names - prev_names)
    removed = sorted(prev_names - cur_names)
    common = sorted(cur_names.intersection(prev_names))

    changed_series = []
    high_count = int(len(duplicate_names_current) + len(duplicate_names_previous))
    for name in common:
        diff = _row_diff(prev_rows[name], cur_rows[name], score_delta_warn=score_delta_warn)
        if diff["severity"] == "none":
            continue
        if diff["severity"] == "high":
            high_count += 1
        changed_series.append({"name": name, **diff})

    status = "no_change"
    if added or removed or changed_series or duplicate_names_current or duplicate_names_previous:
        status = "changed"

    return {
        "status": status,
        "current_count": int(len(cur_rows)),
        "previous_count": int(len(prev_rows)),
        "added_series": added,
        "removed_series": removed,
        "changed_series": changed_series,
        "duplicate_names_current": duplicate_names_current,
        "duplicate_names_previous": duplicate_names_previous,
        "high_severity_count": int(high_count),
        "score_delta_warn": float(score_delta_warn),
    }

File: econark-r/fetchr-R/run/drift_monitor.R
.DRIFT_KEY_FIELDS <- c(
  "status",
  "error",
  "method",
  "disagg_method_used",
  "auto_selection_reason",
  "auto_selection_strategy",
  "profile_name",
  "series_kind",
  "bootstrap_method",
  "indicator_preprocess_mode"
)

.DRIFT_NUMERIC_FIELDS <- c(
  "bootstrap_k_step_selected",
  "auto_selection_score_r2",
  "rho",
  "constraint_benchmark_abs_error"
)

.DRIFT_HIGH_SEVERITY_KEYS <- c("status", "method", "disagg_method_used")

.drift_safe_float <- function(value) {
  if (is.null(value)) return(NA_real_)
  out <- suppressWarnings(as.numeric(value))
  if (!is.finite(out)) return(NA_real_)
  out
}

.drift_safe_scalar <- function(value) {
  if (is.null(value)) return(NA_character_)
  if (is.numeric(value)) {
    if (!is.finite(value)) return(NA_character_)
    return(as.character(value))
  }
  text <- trimws(as.character(value))
  if (!nzchar(text)) return(NA_character_)
  num <- .drift_safe_float(text)
  if (is.finite(num)) return(as.character(num))
  text
}

.drift_row_diff <- function(prev, cur, score_delta_warn = 0.05) {
  changed <- list()
  for (k in .DRIFT_KEY_FIELDS) {
    pv <- .drift_safe_scalar(prev[[k]])
    cv <- .drift_safe_scalar(cur[[k]])
    if (!identical(pv, cv)) changed[[k]] <- list(previous = pv, current = cv)
  }

  numeric_changes <- list()
  score_delta_abs <- 0
  for (k in .DRIFT_NUMERIC_FIELDS) {
    pv <- .drift_safe_float(prev[[k]])
    cv <- .drift_safe_float(cur[[k]])
    if (is.na(pv) && is.na(cv)) next
    if (is.na(pv) || is.na(cv) || abs(cv - pv) > 1e-12) {
      delta <- if (is.na(pv) || is.na(cv)) NA_real_ else as.numeric(cv - pv)
      numeric_changes[[k]] <- list(previous = ifelse(is.na(pv), NULL, pv), current = ifelse(is.na(cv), NULL, cv), delta = ifelse(is.na(delta), NULL, delta))
      if (k == "auto_selection_score_r2" && is.finite(delta)) score_delta_abs <- abs(delta)
    }
  }

  severity <- "none"
  if (any(names(changed) %in% .DRIFT_HIGH_SEVERITY_KEYS)) {
    severity <- "high"
  } else if (length(changed) > 0 || length(numeric_changes) > 0) {
    severity <- "medium"
  }
  if (is.finite(score_delta_abs) && score_delta_abs >= as.numeric(score_delta_warn)) severity <- "high"

  list(
    changed_keys = changed,
    numeric_changes = numeric_changes,
    severity = severity,
    score_delta_abs = as.numeric(score_delta_abs)
  )
}

.drift_rows_by_name <- function(df) {
  rows <- list()
  dups <- character()
  for (i in seq_len(nrow(df))) {
    nm <- as.character(df$name[[i]])
    if (!nzchar(trimws(nm))) next
    if (!is.null(rows[[nm]])) dups <- c(dups, nm)
    row <- as.list(df[i, , drop = FALSE])
    rows[[nm]] <- row
  }
  list(rows = rows, duplicates = sort(unique(dups)))
}

build_interpolation_drift_report <- function(current_summary, previous_summary = NULL, score_delta_warn = 0.05) {
  if (!is.data.frame(current_summary) || !"name" %in% names(current_summary)) {
    stop("current_summary must include name column")
  }
  cur <- current_summary
  cur$name <- as.character(cur$name)
  cur_index <- .drift_rows_by_name(cur)
  cur_rows <- cur_index$rows
  cur_dups <- cur_index$duplicates

  if (is.null(previous_summary) || !is.data.frame(previous_summary) || nrow(previous_summary) == 0L) {
    return(list(
      status = "baseline_initialized",
      current_count = as.integer(length(cur_rows)),
      previous_count = 0L,
      added_series = sort(names(cur_rows)),
      removed_series = list(),
      changed_series = list(),
      duplicate_names_current = as.list(cur_dups),
      duplicate_names_previous = list(),
      high_severity_count = as.integer(length(cur_dups)),
      score_delta_warn = as.numeric(score_delta_warn)
    ))
  }

  if (!"name" %in% names(previous_summary)) {
    return(list(
      status = "previous_summary_invalid",
      current_count = as.integer(length(cur_rows)),
      previous_count = as.integer(nrow(previous_summary)),
      added_series = sort(names(cur_rows)),
      removed_series = list(),
      changed_series = list(),
      duplicate_names_current = as.list(cur_dups),
      duplicate_names_previous = list(),
      high_severity_count = as.integer(length(cur_dups)),
      score_delta_warn = as.numeric(score_delta_warn)
    ))
  }

  prev <- previous_summary
  prev$name <- as.character(prev$name)
  prev_index <- .drift_rows_by_name(prev)
  prev_rows <- prev_index$rows
  prev_dups <- prev_index$duplicates

  cur_names <- sort(names(cur_rows))
  prev_names <- sort(names(prev_rows))
  added <- setdiff(cur_names, prev_names)
  removed <- setdiff(prev_names, cur_names)
  common <- intersect(cur_names, prev_names)

  changed <- list()
  high_count <- length(cur_dups) + length(prev_dups)
  for (nm in common) {
    diff <- .drift_row_diff(prev_rows[[nm]], cur_rows[[nm]], score_delta_warn = score_delta_warn)
    if (diff$severity == "none") next
    if (diff$severity == "high") high_count <- high_count + 1L
    changed[[length(changed) + 1L]] <- c(list(name = nm), diff)
  }

  status <- "no_change"
  if (length(added) > 0 || length(removed) > 0 || length(changed) > 0 || length(cur_dups) > 0 || length(prev_dups) > 0) {
    status <- "changed"
  }

  list(
    status = status,
    current_count = as.integer(length(cur_rows)),
    previous_count = as.integer(length(prev_rows)),
    added_series = as.list(added),
    removed_series = as.list(removed),
    changed_series = changed,
    duplicate_names_current = as.list(cur_dups),
    duplicate_names_previous = as.list(prev_dups),
    high_severity_count = as.integer(high_count),
    score_delta_warn = as.numeric(score_delta_warn)
  )
}
